### PR TITLE
Backend/feat/rental-notifications

### PR DESCRIPTION
**Pull request by prashant601**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/MerchantPushNotification.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  Text: Data.Text
+
+MerchantPushNotification:
+  tableName: merchant_push_notification
+
+  derives: "Generic,Show,ToJSON,FromJSON"
+
+  types:
+    NotificationKey:
+      enum: "FIRST_NOTIFICATION_RIDE_START,SECOND_NOTIFICATION_RIDE_START,RIDE_ALLOCATED,RIDE_ASSIGNED"
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/MerchantPushNotification.yaml:14]: No need to make Enum for this key. Lets make it text type.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  m ()
+sendScheduledRideNotificationsToDriver Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
+  let jobData = jobInfo.jobData
+      merchantOpCityId = jobData.merchantOperatingCityId
+      bookingId = jobData.bookingId
+      driverId = jobData.driverId
+      notificationType = jobData.notificationType
+      notificationKey = jobData.notificationKey
+      onlyIfOffline = jobData.onlyIfOffline
+  smsCfg <- asks (.smsCfg)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs:65]: We don't need this always, lets fetch only when needed.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  booking <- QB.findById bookingId >>= fromMaybeM (BookingDoesNotExist bookingId.getId)
+  driverInfo <- B.runInReplica (QDInfo.getDriverInfo driverId.getId) >>= fromMaybeM DriverInfoNotFound
+  let isNotificationRequired = not onlyIfOffline || (driverInfo.mode /= Just DInfo.ONLINE)
+
+  when (isNotificationRequired && booking.status == DB.TRIP_ASSIGNED) do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs:70]: Maybe keep the status in config itself.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    driver <- B.runInReplica $ QPerson.findById driverId >>= fromMaybeM (PersonDoesNotExist driverId.getId)
+    mobileNumber <- mapM decrypt driver.mobileNumber >>= fromMaybeM (PersonFieldNotPresent "mobileNumber")
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs:72]: Fetch only in case needed.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs [hunk 5]
```diff
@@ -110,6 +119,29 @@ initializeRide merchantId driver booking mbOtpCode enableFrequentLocationUpdates
             "Check the app for more details."
           ]
 
+    pushReminderUpdatesInScheduler DRN.RideRelatedNotificationConfig {..} = do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs:122]: Move it to a separate file in SharedLogic folder.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverInformation/Internal.hs [hunk 6]
```diff
@@ -31,6 +31,13 @@ getDriverInfos ::
 getDriverInfos personKeys = do
   findAllWithKV [Se.Is BeamDI.driverId $ Se.In personKeys]
 
+getDriverInfo ::
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverInformation/Internal.hs:34]: You can use findById

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs (hunk 4 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs:72]: it will be required for CALL notification also

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/MerchantPushNotification.yaml (hunk 1 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/MerchantPushNotification.yaml]: done

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide/Internal.hs [hunk 7]
```diff
@@ -132,6 +135,8 @@ endRideTransaction driverId booking ride mbFareParams mbRiderDetailsId newFarePa
   QDI.updateOnRide False (cast ride.driverId)
   QRide.updateStatus ride.id Ride.COMPLETED
   QRB.updateStatus booking.id SRB.COMPLETED
+  rideRelatedNotificationConfigList <- CRN.findAllByMerchantOperatingCityIdAndTimeDiffEvent booking.merchantOperatingCityId DRN.END_TIME
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide/Internal.hs:138]: Lets add this at the end of the ride end function.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs [hunk 8]
```diff
@@ -85,6 +88,11 @@ initializeRide merchantId driver booking mbOtpCode enableFrequentLocationUpdates
   rideDetails <- buildRideDetails ride driver vehicle
 
   QRB.updateStatus booking.id DBooking.TRIP_ASSIGNED
+
+  when (booking.isScheduled) do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs:92]: Why only for isScheduled booking? It could be other booking as well. Lets move this also inside config.

Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs:92]: And please call this function at the end of functions.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import Lib.Scheduler.JobStorageType.SchedulerType (createJobIn)
+import SharedLogic.Allocator
+
+pushReminderUpdatesInScheduler :: (MonadFlow m, SchedulerFlow r, EsqDBFlow m r, CacheFlow m r) => DB.Booking -> Id DP.Person -> DRN.RideRelatedNotificationConfig -> m ()
+pushReminderUpdatesInScheduler booking driverId DRN.RideRelatedNotificationConfig {..} = do
+  now <- getCurrentTime
+  let eventTime = case timeDiffEvent of
+        DRN.RIDE_ASSIGNED -> booking.updatedAt -- gives the time of TRIP_ASSIGNED
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:31]: Please use, ride.created_at.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        DRN.PICKUP_TIME -> booking.startTime -- scheduled time
+        _ -> now -- this will not get scheduled
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:33]: What about ride start time and end time?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  let currentTimeDiffFromScheduledTime = diffUTCTime eventTime now
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:34]: Did you test it? It should be `diffUTCTime now eventTime`. But this logic is incorrect.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs [hunk 12]
```diff
@@ -219,6 +224,9 @@ rideAssignedReqHandler req = do
         Nothing -> "normal"
   incrementRideCreatedRequestCount booking.merchantId.getId booking.merchantOperatingCityId.getId category
   _ <- QRB.updateStatus booking.id DRB.TRIP_ASSIGNED
+  when (booking.isScheduled) do
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs:227]: Same as driver.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs [hunk 13]
```diff
@@ -311,6 +320,8 @@ rideStartedReqHandler ValidatedRideStartedReq {..} = do
   triggerRideStartedEvent RideEventData {ride = updRideForStartReq, personId = booking.riderId, merchantId = booking.merchantId}
   _ <- QRide.updateMultiple updRideForStartReq.id updRideForStartReq
   _ <- QPFS.updateStatus booking.riderId DPFS.RIDE_STARTED {rideId = ride.id, bookingId = booking.id, trackingUrl = ride.trackingUrl, driverLocation = Nothing}
+  rideRelatedNotificationConfigList <- CRRN.findAllByMerchantOperatingCityIdAndTimeDiffEvent booking.merchantOperatingCityId DRN.START_TIME
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs:323]: Also add this at driver side.

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+import SharedLogic.JobScheduler
+
+pushReminderUpdatesInScheduler :: (MonadFlow m, SchedulerFlow r, EsqDBFlow m r, CacheFlow m r) => DB.Booking -> DRN.RideRelatedNotificationConfig -> m ()
+pushReminderUpdatesInScheduler booking DRN.RideRelatedNotificationConfig {..} = do
+  now <- getCurrentTime
+  let eventTime = case timeDiffEvent of
+        DRN.RIDE_ASSIGNED -> booking.updatedAt -- gives the time of TRIP_ASSIGNED
+        DRN.PICKUP_TIME -> booking.startTime -- scheduled time
+        _ -> now -- this will not get scheduled
+  let currentTimeDiffFromScheduledTime = diffUTCTime eventTime now
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/ScheduledNotifications.hs:32]: Resolve for driver and rider both.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/Notifications.hs [hunk 15]
```diff
@@ -896,3 +896,42 @@ cityFallback :: Maybe Version.Version -> Id DMOC.MerchantOperatingCity -> Id DMO
...
+  Id DMOC.MerchantOperatingCity ->
+  Id Person ->
+  Maybe FCM.FCMRecipientToken ->
+  ReminderReq ->
+  m ()
+notifyScheduledRideReminderToDriver merchantOpCityId personId mbDeviceToken entityData = do
+  transporterConfig <- findByMerchantOpCityId merchantOpCityId (Just (DriverId (cast personId))) >>= fromMaybeM (TransporterConfigNotFound merchantOpCityId.getId)
+  FCM.notifyPersonWithPriority transporterConfig.fcmConfig (Just FCM.HIGH) False notificationData (FCMNotificationRecipient personId.getId mbDeviceToken) EulerHS.Prelude.id
+  where
+    notifType = FCM.SCHEDULED_RIDE_NOTIFICATION
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/Notifications.hs:920]: Move it merchant_push_notification.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+        sendOverlay merchantOpCityId driver $ mkOverlayReq overlay
+      SMS -> do
+        smsCfg <- asks (.smsCfg)
+        let sender = smsCfg.sender
+        messageKey <- A.decode (A.encode notificationKey) & fromMaybeM (InvalidRequest "Invalid message key for SMS")
+        merchantMessage <- CMM.findByMerchantOpCityIdAndMessageKey merchantOpCityId messageKey >>= fromMaybeM (MerchantMessageNotFound merchantOpCityId.getId notificationKey)
+        Sms.sendSMS driver.merchantId merchantOpCityId (Sms.SendSMSReq merchantMessage.message phoneNumber sender) >>= Sms.checkSmsResult
+      _ -> pure () -- WHATSAPP or Other Notifications can be implemented here
+  where
+    generateReminderReq notifTitle notifBody booking = do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs:94]: Make it generic.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import SharedLogic.Allocator
+
+pushReminderUpdatesInScheduler :: (MonadFlow m, SchedulerFlow r, EsqDBFlow m r, CacheFlow m r) => DB.Booking -> DR.Ride -> UTCTime -> Id DP.Person -> DRN.RideRelatedNotificationConfig -> m ()
+pushReminderUpdatesInScheduler booking ride now driverId DRN.RideRelatedNotificationConfig {..} = do
+  let eventTime = case timeDiffEvent of
+        DRN.RIDE_ASSIGNED -> ride.createdAt -- gives the time of TRIP_ASSIGNED
+        DRN.PICKUP_TIME -> booking.startTime -- scheduled time
+        DRN.START_TIME -> fromMaybe now ride.tripStartTime
+        DRN.END_TIME -> fromMaybe now ride.tripEndTime
+  let currentTimeDiffFromScheduledTime = diffUTCTime eventTime now
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:35]: Rename it to `currentTimeDiffFromEventTime`.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+      toSchedule = if onScheduledBooking then booking.isScheduled else True
+  when (toSchedule && currentTimeDiffFromScheduledTime >= timeDiff) do
+    let dfCalculationJobTs = timeDiff + 2
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:38]: `max 2 (currentTimeDiffFromScheduledTime - timeDiff)`

Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:38]: data EventTime = PreEvent | PostEvent | OnEvent

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import Lib.Scheduler.JobStorageType.SchedulerType (createJobIn)
+import SharedLogic.Allocator
+
+pushReminderUpdatesInScheduler :: (MonadFlow m, SchedulerFlow r, EsqDBFlow m r, CacheFlow m r) => DB.Booking -> DR.Ride -> UTCTime -> Id DP.Person -> DRN.RideRelatedNotificationConfig -> m ()
+pushReminderUpdatesInScheduler booking ride now driverId DRN.RideRelatedNotificationConfig {..} = do
+  let eventTime = case timeDiffEvent of
+        DRN.RIDE_ASSIGNED -> ride.createdAt -- gives the time of TRIP_ASSIGNED
+        DRN.PICKUP_TIME -> booking.startTime -- scheduled time
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:32]: Handle pick up time event for customer.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs [hunk 20]
```diff
@@ -102,6 +105,10 @@ initializeRide merchantId driver booking mbOtpCode enableFrequentLocationUpdates
   fork "DriverScoreEventHandler OnNewRideAssigned" $
     DS.driverScoreEventHandler booking.merchantOperatingCityId DST.OnNewRideAssigned {merchantId = merchantId, driverId = ride.driverId, currency = ride.currency}
 
+  now <- getCurrentTime
+  rideRelatedNotificationConfigList <- SCRRNC.findAllByMerchantOperatingCityIdAndTimeDiffEvent booking.merchantOperatingCityId DRN.RIDE_ASSIGNED
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs:109]: Also add PICKUP_TIME.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/OnDemand/Transformer/OnUpdate.hs [hunk 21]
```diff
@@ -243,3 +243,21 @@ buildOnUpdateReqOrderV2 req' mbFarePolicy becknConfig = case req' of
           orderCreatedAt = Just booking.createdAt,
           orderUpdatedAt = Just booking.updatedAt
         }
+  OU.TollCrossedBuildReq OU.DTollCrossedBuildReq {..} -> do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/OnDemand/Transformer/OnUpdate.hs:246]: This is for?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs [hunk 22]
```diff
@@ -524,6 +528,11 @@ calculateFinalValuesForCorrectDistanceCalculations handle booking ride mbMaxDist
       SWC.incrementByValue (round distanceDiff) weeklyExtraKmsKey SlidingWindowOptions {period = 7, periodType = Days}
       pure (Just dailyExtraKms, Just weeklyExtraKms)
 
+    notifyDriverOnExtraKmsLimitExceed = do
+      driver <- QP.findById ride.driverId >>= fromMaybeM (PersonNotFound ride.driverId.getId)
+      overlay <- CMP.findByMerchantOpCityIdPNKeyLangaugeUdf booking.merchantOperatingCityId "EXTRA_KMS_LIMIT_EXCEEDED" (fromMaybe ENGLISH driver.language) Nothing >>= fromMaybeM (OverlayKeyNotFound "EXTRA_KMS_LIMIT_EXCEEDED")
+      fork "Send Extra Kms Limit Exceeded Overlay" $ TN.sendOverlay booking.merchantOperatingCityId driver $ TN.mkOverlayReq overlay
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs:534]: Instead of fork here, please add fork at start of this function.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CallBAP.hs [hunk 23]
```diff
@@ -774,6 +775,43 @@ sendQuoteRepetitionUpdateToBAP booking ride newBookingId cancellationSource driv
...
+    HasFlowEnv m r '["ondcTokenHashMap" ::: HMS.HashMap KeyConfig TokenConfig],
+    HasFlowEnv m r '["internalEndPointHashMap" ::: HMS.HashMap BaseUrl BaseUrl],
+    HasFlowEnv m r '["kafkaProducerTools" ::: KafkaProducerTools]
+  ) =>
+  Maybe DRB.Booking ->
+  Maybe SRide.Ride ->
+  DP.Person ->
+  DVeh.Vehicle ->
+  m ()
+sendTollCrossedUpdateToBAP (Just booking) (Just ride) driver vehicle = do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CallBAP.hs:794]: Can we make it part of separate PR>

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+pushReminderUpdatesInScheduler :: (MonadFlow m, SchedulerFlow r, EsqDBFlow m r, CacheFlow m r) => DB.Booking -> DR.Ride -> UTCTime -> Id DP.Person -> DRN.RideRelatedNotificationConfig -> m ()
+pushReminderUpdatesInScheduler booking ride now driverId DRN.RideRelatedNotificationConfig {..} = do
+  let eventAt = case timeDiffEvent of
+        DRN.RIDE_ASSIGNED -> ride.createdAt -- gives the time of TRIP_ASSIGNED
+        DRN.PICKUP_TIME -> booking.startTime -- scheduled time
+        DRN.START_TIME -> fromMaybe now ride.tripStartTime
+        DRN.END_TIME -> fromMaybe now ride.tripEndTime
+  let currentTimeDiffFromEventTime = diffUTCTime eventAt now
+      toSchedule = if onScheduledBooking then booking.isScheduled else True
+  when (toSchedule && currentTimeDiffFromEventTime >= timeDiff) do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:37]: currentTimeDiffFromEventTime >= timeDiff

why this check, it can fail in case of PostEvent.

**Review by hkmangla — DISMISSED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/OnDemand/Transformer/OnUpdate.hs (hunk 21 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/OnDemand/Transformer/OnUpdate.hs:246]: these are toll related changes

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs (hunk 24 above)
Comment (prashant601) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/ScheduledNotifications.hs:37]: in Case of preEvent , we can't schedule for the time which is already passed, this check was for that.

we can have it as 
`when (toSchedule && ( eventTime /= DRN.PreEvent || currentTimeDiffFromEventTime >= timeDiff) )`

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs [hunk 25]
```diff
@@ -443,6 +466,10 @@ validateRequest = \case
     when (ride.status == DRide.COMPLETED) $ throwError $ RideInvalidStatus "Can't edit the destination of a completed ride."
     bookingUpdateRequest <- runInReplica $ QBUR.findById bookingUpdateRequestId >>= fromMaybeM (InternalError $ "BookingUpdateRequest not found with Id:-" <> bookingUpdateRequestId.getId)
     return $ OUValidatedEditDestConfirmUpdateReq ValidatedEditDestConfirmUpdateReq {..}
+  OUTollCrossedEventReq TollCrossedEventReq {..} -> do
+    booking <- QEBooking.findByTransactionId transactionId >>= fromMaybeM (BookingDoesNotExist $ "transactionId - " <> transactionId)
```
Comment (khuzema786) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:470]: Let's not throw error here

**Review by khuzema786 — APPROVED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CallBAP.hs (hunk 23 above)
Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CallBAP.hs:794]: :(
